### ISSUE TITLE
LiveTable and plot don't update live

### ISSUE THREAD

**Issue by brunoseivam** [NSLS-II/Bug-Reports#131]
I am using Jupyter on `xf17id1-ca1` to serve notebooks to FMX scientists. It works fine but the plots and the LiveTable don't seem to update live during a scan, only all at once in the end. If that's relevant, `jupyter` is run as the user `notebook`, which doesn't have `DISPLAY` set. In [`startup/01-bluesky.py`](https://github.com/NSLS-II-FMX/profile_collection/blob/master/startup/01-bluesky.py) I conditionally call `install_qt_kicker` like so:

```
# Only install_qt_kicker if DISPLAY is set
import os
if "DISPLAY" in os.environ:
    from bluesky.utils import install_qt_kicker
    # The following line allows bluesky and pyqt4 GUIs to play nicely together:
    install_qt_kicker()
```

I am calling `%matplotlib notebook`.

**Comment by danielballan** [comment]
A couple things here:
- It's technically possible to do data collection from a notebook, but we don't currently recommend it. It adds more layers (possible points of failure) between the user and the motors, increasing the potential for bad things to happen. In my personal opinion, this is something we should and probably will support in the medium-future, but the present consensus (which, to be clear, I agree with -- for now) is that you shouldn't move hardware from a notebook.
- We don't currently have a way to live-update `LivePlot` in the notebook. We think it is technically possible, but we haven't worked out the details yet.
- But `LiveTable` should be updating. It has nothing to do with $DISPLAY or the qt kicker, and I've seen it work in notebooks just fine. Are you sure `LiveTable` is failing to update? What happens if you use `LiveTable` without `LivePlot`?

**Comment by brunoseivam** [comment]
I tried:

```
RE(count([cam_mono], 5, 1), LiveTable([cam_mono]))
```

On the notebook, `LiveTable` prints the header immediately and then the rest of the body only after 5 seconds. Using `ipython` on the command line I get updates every second. Both of them use the same [profile](https://github.com/NSLS-II-FMX/profile_collection). 

Regarding moving motors from the notebook, do you have a worst case scenario? Would losing the connection from the browser to the server machine cause motors to move uncontrollably? If that's the case, wouldn't that happen also if the user is connecting via ssh and the connection drops? Or is it more of a "things might not work properly"?

**Comment by hhslepicka** [comment]
@danielballan I can see the same at LIX.

**Comment by danielballan** [comment]
Yes, the scenario I have in mind is a dropped connection. We would say that users shouldn't be moving motors of ssh either, given the potential to break things.

I get row-by-row updates when I run `LiveTable` in the notebook. Can you paste a reproducible example?

**Comment by brunoseivam** [comment]
So, the recommended way of moving motors right now is having a local instance of bluesky on every workstation?

To reproduce:

Open `xf17id1-ca1.cs.nsls2.local:17000` in a browser (box64 has access). Create a notebook and run `RE(count([cam_mono], 5, 1), LiveTable([cam_mono]))`

**Comment by danielballan** [comment]
SSH connections between local machines are relatively safe, so I would not say that it's necessary to have separate local installations on every workstation. Remote SSH connections or notebooks (which hop through many machines) are less stable.

**Comment by danielballan** [comment]
OK, I can reproduce. I think a well-placed `sys.stdout.flush()` might fix this. Care to try editing `bluesky/callbacks/core.py` and put in a PR?

**Comment by tacaswell** [comment]
This is a bad idea (as @danielballan enumerated above).

I'll happily review and merge a `install_nb_kicker` function.

**Comment by brunoseivam** [comment]
@danielballan The problem with `LiveTable` doesn't seem to be trivial. `flush` doesn't work. Looks like it is something `bluesky` related, though, since `ophyd`'s `movr` prints fine, for example. Something is buffering what it is not supposed to. 

Wireshark traces on both the client and the server machines show that the body of the `LiveTable` is sent all at once in a single packet, so I would rule out a network device doing the buffering.

**Comment by danielballan** [comment]
Just to be sure we're on the same page, did you try putting a call to flush inside `LiveTable.event`?

**Comment by brunoseivam** [comment]
Yes, and inside `LiveTable._print`, and `print(..., flush=True)`, and wrapping `sys.stdout` so it always flushes after writes.

**Comment by danielballan** [comment]
Then you definitely covered the bases. There is probably an IPython-specific hook we need to hit here.
